### ISSUE TITLE
Allow automatic conversion of GR(1) compatible specs

### ISSUE THREAD

**Issue by slivingston** [tulip-control/tulip-control#35]
(Copied from https://sourceforge.net/p/tulip-control/feature-requests/13/ with creation date 2012-05-30, by @murrayrm but assigned to no one.)

We should create a utility function that will automatically take specification components that can be converted to GR(1) compatible specifications and create the appropriate auxiliary variables, system dynamics and specifications. Some of this functionality is built into JTLV, but by pulling it out and putting it into TuLiP, we can use it with gr1c and other tools.

COMMENTS copied from https://sourceforge.net/p/tulip-control/feature-requests/13/:

2012-06-07 by @necozay: I am playing with the past fragment in JTLV and looking into how conversions work for certain fragments of LTL. I can work on this feature request over the summer if noone else signs up for it. Or, I will be happy to help whoever ends up working on this.

In addition to this it would be nice have some standardization of inputs. Currently we have two different syntax for GRSpec, one compatible with JTLV , one with gr1c. The conversion in between these different formats are not completely in there. Might be something useful. Nick may also hep out with this since he will be using the whole LTL (with nusmv and spin), so some standardization, error handling would be needed for his work as well.

**Comment by slivingston** [comment]
I bumped the milestone for this issue because the second action item [described above](https://github.com/tulip-control/tulip-control/issues/35#issuecomment-48450038) remains. I expect that making a general and robust implementation will be a substantial effort. I recommend that the first action item [described above](https://github.com/tulip-control/tulip-control/issues/35#issuecomment-48450038) is considered done. Future desires can be expressed by re-opening this issue or opening an issue about the specific desired re-write rule.

Observe that the OP can be interpreted as a request that is much easier than the aforementioned action item and that it is essentially done as of ab6476021812b4b55fd7970f57da78f245254130. How about we close this issue after creating a small demonstration, e.g., under the directory `examples/developer/`? We could then open a separate issue about a routine for attempting to construct a `GRSpec` object automatically from an instance of `LTL`, or less specifically, an issue about automatic splitting of an arbitrary LTL formula into init, safety, and liveness ("progress") components [as described above](https://github.com/tulip-control/tulip-control/issues/35#issuecomment-48450038).

**Comment by johnyf** [comment]
I agree that this issue can be closed, considering item 1 [above](https://github.com/tulip-control/tulip-control/issues/35#issuecomment-48450038) complete.

Item 2 can be deferred. The suggested approach is syntactic and enumerative (in the sense of manipulating the syntax tree). This approach will be inefficient, cumbersome and very error prone to implement, and will still be limited to the few cases recognized by item 1. For this reason, I recommend cancelling item 2.

Instead, I suggest that item 2 be replaced by:

2. if a temporal property is described by an LTL formula not in the syntactic fragment known as GR(1), then:
    1. check if the property can be described by a deterministic Buchi automaton
    2. if No.1 is true, then construct such an automaton
    3. return a BDD representation of the deterministic automaton (which is in the GR(1) fragment, e.g., express the BDDs as formulae containing "next").

The determinization can be performed symbolically, as described and implemented in [Morgenstern 2011](http://arxiv.org/abs/1102.4119#) (the implementation is in F#, closed-source, and failed to run in my case). The details are described in [Schneider, Verification of reactive systems, Springer](http://www.springer.com/us/book/9783540002963). The implementation requires a symbolic implementation of the Miyano-Hayashi breakpoint construction (from what I recall), and other transformations. This task can take a substantial amount of time, in the order of days or weeks.

To answer also the question [about persistence `<>[]p`](https://github.com/tulip-control/tulip-control/issues/35#issuecomment-48491591), there does not exist a deterministic Buchi automaton that describes this property, so it cannot be expressed in the GR(1) fragment. This corrects my speculation [above](https://github.com/tulip-control/tulip-control/issues/35#issuecomment-48505700).

The equivalence should hold (formal proof required here) only if all variables are existentially quantified (a "closed system"). Not in the presence of universal quantification, unless the realizability of the temporal property `<>[]p` is completely independent of any universal quantification. If universal quantification interacts in any way with the property `<>[]p`, then the equivalent non-deterministic BA cannot be treated as a deterministic one, because it's runs are over an alphabet that involves universally quantified variables.

The explanation is that the variable `aux` represents the Buchi automaton's node. If universal quantification interferes with the BA, then `aux` should be considered as an environment variable hidden from the system (even if treated as a system variable, it is affected by universal quantification). Treating it as a visible variable (the only option in GR(1)) describes a deterministic BA that looks similar to the nondeterministic BA, except that the BA's current node (`aux`) is part of the alphabet, and thus annotates edges (is "visible"). This is what can cause confusion.

Note: item 2 refers to splitting arbitrary LTL formulae, as commented by @slivingston. Splitting of GR(1) formulae has been [implemented](https://github.com/tulip-control/tulip-control/blob/a5fd2bfb715df726905bf301a020ffac684bb35f/tulip/spec/gr1_fragment.py#L191).
